libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/ChargingStationsPreserver.kt

```diff
@@ -0,0 +1,64 @@
+package com.mapbox.navigation.core.reroute
+
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.routeoptions.isEVRoute
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+internal suspend fun preserveChargingStationsFromOldRoutes(
+    originalRoute: NavigationRoute,
+    newRoutes: List<NavigationRoute>
+): List<NavigationRoute> = withContext(Dispatchers.Default) {
+    if (!originalRoute.routeOptions.isEVRoute()) {
+        return@withContext newRoutes
+    }
+    if (newRoutes.first().origin == RouterOrigin.Offboard) {
+        return@withContext newRoutes
+    }
+    val newPrimaryRoute = newRoutes.first()
+    val rerouteRouteOptions = newPrimaryRoute.routeOptions
+    val chargingStationsIds = rerouteRouteOptions.getUnrecognisedParameter("waypoints.charging_station_id")
+        ?: return@withContext newRoutes
```

Comment:
Why do you return here? It's entirely possible that not all of these parameters are filled. The only mandatory parameter is "power".

Reply:
agree, done